File: DFS/695.Max-Area-of-Island/695.Max-Area-of-Island.ts
function maxAreaOfIsland(grid: number[][]): number {
  let currentSize = 0;
  let ret = 0;
  const m = grid.length;
  const n = grid[0].length;

  for (let i = 0; i < m; i++) {
    for (let j = 0; j < n; j++) {
      if (grid[i][j] === 0 || grid[i][j] === 2) {
        continue;
      }
      currentSize = 0;
      dfs(i, j);
      ret = Math.max(currentSize, ret);
    }
  }

  return ret;

  function dfs(i: number, j: number): void {
    if (i < 0 || i >= m || j < 0 || j >= n) {
      return;
    }

    if (grid[i][j] === 0 || grid[i][j] === 2) {
      return;
    }

    grid[i][j] = 2;
    currentSize += 1;

    dfs(i + 1, j);
    dfs(i - 1, j);
    dfs(i, j + 1);
    dfs(i, j - 1);
  }
}

File: DFS/695.Max-Area-of-Island/695.Max-Area-of-Island.py
class Solution:
    def maxAreaOfIsland(self, grid: List[List[int]]) -> int:
        m = len(grid[0])
        n = len(grid)
        def dfs(i: int, j: int) -> int:
            if i < 0 or i >= n or j < 0 or j >= m:
                return 0
            
            if grid[i][j] == 0 or grid[i][j] == 2:
                return 0
            
            grid[i][j] = 2

            sum = 1
            direction = [[-1, 0], [1, 0], [0, 1], [0, -1]]
            for [x, y] in direction:
                sum += dfs(i + x, j + y)
            return sum
        
        result = 0
        for i in range(n):
            for j in range(m):
                if grid[i][j] == 1:
                    result = max(result, dfs(i, j))
        
        return result
